encoders/firebase-decoders-json/src/test/java/com/google/firebase/decoders/json/JsonDataDecoderBuilderContextTest.java

```diff
@@ -0,0 +1,137 @@
+import com.google.firebase.decoders.ObjectDecoder;
+import com.google.firebase.decoders.ObjectDecoderContext;
+import com.google.firebase.decoders.Safe;
+import com.google.firebase.decoders.TypeCreator;
+import com.google.firebase.decoders.TypeToken;
+import com.google.firebase.decoders.json.JsonDataDecoderBuilderContext;
+import com.google.firebase.encoders.FieldDescriptor;
+import com.google.firebase.encoders.annotations.Encodable;
+
+import static com.google.common.truth.Truth.assertThat;
+import static org.junit.Assert.assertThrows;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+import org.robolectric.RobolectricTestRunner;
+/**
+ * TODO: TypeToken support TypeVariable, inorder to support decode nested Generic type
+ * class Foo<T> {
+ *   SubFoo<T>
+ * }
+ */
+@RunWith(RobolectricTestRunner.class)
+public class JsonDataDecoderBuilderContextTest {
+
+  static class Foo<T> {
+    int a;
+    boolean b;
+    T t;
+    SubFoo<String> subFoo;
+
+    public Foo(int a, boolean b, T t, SubFoo<String> subFoo) {
+      this.a = a;
+      this.b = b;
+      this.t = t;
+      this.subFoo = subFoo;
+    }
+  }
+
+  static class SubFoo<T> {
+    int a;
+    boolean b;
+    T t;
+
+    public SubFoo(int a, boolean b, T t) {
+      this.a = a;
+      this.b = b;
+      this.t = t;
+    }
+  }
+
+  static class FooObjectDecoder<T> implements ObjectDecoder<Foo<T>> {
+    @NonNull
+    @Override
+    public TypeCreator<Foo<T>> decode(@NonNull ObjectDecoderContext<Foo<T>> ctx) {
+      FieldRef.Primitive<Integer> aField = ctx.decodeInteger(FieldDescriptor.of("a"));
+      FieldRef.Primitive<Boolean> bField = ctx.decodeBoolean(FieldDescriptor.of("b"));
+      FieldRef.Boxed<T> tField = ctx.decode(FieldDescriptor.of("t"), (TypeToken.ClassToken) ctx.getTypeArguments().at(0));
+      FieldRef.Boxed<SubFoo<String>> subFooField = ctx.decode(FieldDescriptor.of("subFoo"), TypeToken.of(new Safe<SubFoo<String>>(){}));
+
+      return (creationCtx ->
+              new Foo<T>(
+                      creationCtx.getInteger(aField),
+                      creationCtx.getBoolean(bField),
+                      (T) creationCtx.get(tField),
+                      (SubFoo<String>) creationCtx.get(subFooField)
+                      )
+      );
+    }
+  }
+
+  static class SubFooObjectDecoder<T> implements ObjectDecoder<SubFoo<T>> {
+
+    @NonNull
+    @Override
+    public TypeCreator<SubFoo<T>> decode(@NonNull ObjectDecoderContext<SubFoo<T>> ctx) {
+      FieldRef.Primitive<Integer> aField = ctx.decodeInteger(FieldDescriptor.of("a"));
+      FieldRef.Primitive<Boolean> bField = ctx.decodeBoolean(FieldDescriptor.of("b"));
+      FieldRef.Boxed<T> tField = ctx.decode(FieldDescriptor.of("t"), (TypeToken.ClassToken) ctx.getTypeArguments().at(0));
```

Comment:
Did you have to cast to `(TypeToken.ClassToken)` here? I think the types should align correctly without the cast.